### DISCUSSION TITLE
Devicescript and wasm3 integration

### DISCUSSION THREAD

**Discussion by Sebight** [microsoft/devicescript#691] [Q&A]
Hello, I am looking around and wondering whether it's possible to run Devicescript's compiled bytecode on the **wasm3** interpreter. I suppose the question is more towards: **Does Devicescript generate a proprietary WASM bytecode or rather a general one, that can be run on various interpreters?**

Thanks a lot!

**Comment by pelikhan** [comment]
@mmoskal

**Comment by mmoskal** [comment]
DeviceScript uses [custom bytecode format](https://microsoft.github.io/devicescript/language/bytecode). The interpreter for that format can be compiled for any version of Wasm (which can then be executed by a variety of Wasm runtimes).

See also [Runtime implementation](https://microsoft.github.io/devicescript/language/runtime)

**Reply by Sebight** [reply]
Got that, thanks. So it is coupled with WASM and it cannot be used standalone, correct?

I am asking because I found out that WASM has a RAM page of 64kB and we won't be able to afford that. :/

**Reply by mmoskal** [reply]
The interpreter for DeviceScript bytecode can be compiled for many devices, see https://microsoft.github.io/devicescript/devices

None of these involve Wasm.

**Reply by Sebight** [reply]
Hello, thanks for the answer. As I understood, this is the interpreter: https://github.com/microsoft/devicescript-esp32 that be compiled for ESP32.

Would there be any documentation for the C++ interaction with the interpreter and so on? The only thing I found was this, but that does not seem to include these (or I can't search). https://microsoft.github.io/devicescript/language/runtime

Thank you!
